File: StandardCyborgAlgorithmsTestbed/TestViewControllers/NodeToNodeLineNode.m
//
//  NodeToNodeLineNode.m
//  StandardCyborgGeometryTestbed
//
//  Created by Aaron Thompson on 4/6/19.
//  Copyright © 2019 Standard Cyborg. All rights reserved.
//

#import "NodeToNodeLineNode.h"

static CGFloat SCNVector3Distance(SCNVector3 a, SCNVector3 b)
{
    CGFloat dx = a.x - b.x;
    CGFloat dy = a.y - b.y;
    CGFloat dz = a.z - b.z;
    
    return sqrt(dx * dx + dy * dy + dz * dz);
}

static SCNVector3 SCNVector3Midpoint(SCNVector3 a, SCNVector3 b)
{
    CGFloat dx = b.x - a.x;
    CGFloat dy = b.y - a.y;
    CGFloat dz = b.z - a.z;
    
    return SCNVector3Make(dx / 2.0 + a.x, dy / 2.0 + a.y, dz / 2.0 + a.z);
}

@implementation NodeToNodeLineNode {
    SCNNode *_cylinderNode;
}

- (instancetype)initWithStartNode:(SCNNode *)startNode
                          endNode:(SCNNode *)endNode
                        thickness:(CGFloat)thickness
                            color:(UIColor *)color
{
    self = [super init];
    if (self) {
        __weak SCNNode *weakStartNode = startNode;
        __weak SCNNode *weakEndNode = endNode;
        
        __weak SCNCylinder *cylinder = [SCNCylinder cylinderWithRadius:thickness height:1];
        cylinder.firstMaterial.diffuse.contents = color;
        _cylinderNode = [SCNNode nodeWithGeometry:cylinder];
        _cylinderNode.eulerAngles = SCNVector3Make(M_PI_2, 0, 0);
        [self addChildNode:_cylinderNode];
        
        SCNConstraint *positionConstraint = [SCNTransformConstraint positionConstraintInWorldSpace:NO withBlock:^SCNVector3(SCNNode *node, SCNVector3 position) {
            __strong SCNNode *strongStartNode = weakStartNode;
            __strong SCNNode *strongEndNode = weakEndNode;
            if (strongStartNode == nil || strongEndNode == nil) { return SCNVector3Zero; }
            
            // Update the cylinder's height while we're at it
            cylinder.height = SCNVector3Distance(startNode.position, endNode.position);
            
            SCNVector3 start = strongStartNode.position;
            SCNVector3 end = strongEndNode.position;
            SCNVector3 middle = SCNVector3Midpoint(start, end);
            
            return middle;
        }];
        positionConstraint.incremental = NO;
        
        SCNConstraint *lookAtConstraint = [SCNLookAtConstraint lookAtConstraintWithTarget:weakEndNode];
        lookAtConstraint.incremental = NO;
        
        [self setConstraints:@[positionConstraint, lookAtConstraint]];
    }
    return self;
}

@end

File: StandardCyborgUI/StandardCyborgUI/NodeToNodeLineNode.swift
//
//  NodeToNodeLineNode.swift
//  StandardCyborgUI
//
//  Created by Aaron Thompson on 7/15/19.
//  Copyright © 2019 Standard Cyborg. All rights reserved.
//

import SceneKit

@objc public class NodeToNodeLineNode: SCNNode {
    
    private let _cylinderNode: SCNNode
    
    @objc public init(startNode: SCNNode,
                      endNode: SCNNode,
                      thickness: CGFloat,
                      color: UIColor)
    {
        let cylinder = SCNCylinder(radius: thickness, height: 1)
        cylinder.firstMaterial?.diffuse.contents = color
        _cylinderNode = SCNNode(geometry: cylinder)
        _cylinderNode.eulerAngles = SCNVector3(x: Float.pi / 2.0, y: 0, z: 0)
        
        super.init()
        
        addChildNode(_cylinderNode)
        
        let positionConstraint = SCNTransformConstraint.positionConstraint(inWorldSpace: false) {
            [weak startNode, weak endNode, unowned cylinder]
            (node: SCNNode, position: SCNVector3) -> SCNVector3 in
            guard let startNode = startNode, let endNode = endNode else { return SCNVector3Zero }
            
            // Update the cylinder's height while we're at it
            cylinder.height = CGFloat(SCNVector3Distance(startNode.position, endNode.position))
            
            let start = startNode.position
            let end = endNode.position
            let middle = SCNVector3Midpoint(start, end)
            
            return middle
        }
        
        let lookAtConstraint = SCNLookAtConstraint(target: endNode)
        
        positionConstraint.isIncremental = false
        lookAtConstraint.isIncremental = false
        
        constraints = [positionConstraint, lookAtConstraint]
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}

public func SCNVector3Distance(_ a: SCNVector3, _ b: SCNVector3) -> Float
{
    let dx: Float = a.x - b.x
    let dy: Float = a.y - b.y
    let dz: Float = a.z - b.z
    
    return sqrtf(dx * dx + dy * dy + dz * dz)
}

public func SCNVector3Midpoint(_ a: SCNVector3, _ b: SCNVector3) -> SCNVector3
{
    let dx: Float = b.x - a.x
    let dy: Float = b.y - a.y
    let dz: Float = b.z - a.z
    
    return SCNVector3(x: dx / 2.0 + a.x,
                      y: dy / 2.0 + a.y,
                      z: dz / 2.0 + a.z)
}
